components/conversation/src/Conversation.svelte

```diff
@@ -145,30 +145,62 @@
   }
 
   //#region Contacts
-  $: {
-    if (participants.length) {
-      setContacts(participants);
+  let contacts: any = null;
+  let contact_query: ContactSearchQuery;
+  $: contact_query = {
+    component_id: id,
+    access_token,
+    query: "",
+  };
+
+  $: (async () => {
+    if (!contacts && conversation) {
+      await getContacts(conversation);
+    }
+  })();
+
+  async function getContacts(conversation: Conversation) {
+    const fromParticipants =
+      conversation.messages?.reduce<Record<string, Participant>>(
+        (participants, message) => {
+          const participant: Participant = message.from[0];
+          participants[participant.email] = participant;
+          return participants;
+        },
+        {},
+      ) || {};
+    const fromParticipantsArray =
+      Array.from(Object.values(fromParticipants)) || [];
+
+    for (const participant of fromParticipantsArray) {
+      const participantEmail = participant.email;
+      if (!contacts) {
+        contacts = {};
+      }
+
+      if (participantEmail && !contacts[participantEmail]) {
+        contacts[participantEmail] = await getContact(participant);
+      }
     }
   }
 
-  function setContacts(participants: Participant[]) {
-    participants
-      .filter((participant) => participant.contact === undefined) // only the ones that aren't yet hydrated
-      .forEach((participant) => {
-        fetchContactsByQuery({
-          component_id: id,
-          // contact_id: participant.id
-          query: `?email=${participant.email}`,
-          access_token,
-        }).then((contacts) => {
-          if (contacts.length) {
-            participant.contact = contacts[0];
-          } else {
-            participant.contact = null;
-          }
-        });
-      });
+  /*
+    Fetches contact for ContactImage component
+  */
+  async function getContact(participant: Participant) {
+    contact_query["query"] = `?email=${participant.email}`;
+
+    if (id) {
+      let contact = $ContactStore[JSON.stringify(contact_query)];
+      if (!contact) {
+        contact = await ContactStore.addContact(contact_query);
+      }
+      return contact[0] ?? participant;
+    } else {
+      return participant;
+    }
   }
+  // #endregion get contact for ContactImage
```

Comment:
this endregion doesn't have a start!

Reply:
Oh, #187 is the start. Maybe I can add an additional `// #region` there.